File: src/molgr/fallback/stages/preprocess.py
"""Early structure preparation and validation stages for fallback."""

from __future__ import annotations

import itertools
from typing import List, Optional, Tuple, cast

from openbabel import openbabel as ob
from openbabel import pybel

from molgr.fallback.utils import smarts
from molgr.fallback.utils.electrons import (
    get_unpaired_electron_count,
    has_unresolved_two_electron_center,
    set_lone_pair_count,
    set_unpaired_electron_count,
    set_unresolved_two_electron_center,
)


def validate_omol(
    omol: pybel.Molecule, total_charge: int = 0, total_radical_electrons: int = 0
) -> bool:
    """Resolve deferred two-electron centers against the global radical budget.

    Fixed atoms contribute only their real unpaired-electron counts; active lone
    pairs do not enter the spin budget. If ``T - R`` is an admissible even number,
    the first required unresolved centers become triplets ``(2, 0)`` and the rest
    singlets ``(0, 1)`` in stable atom order. An impossible budget returns false
    without modifying any marker or occupancy field.
    """

    formal_charge_sum = sum(cast(ob.OBAtom, atom.OBAtom).GetFormalCharge() for atom in omol.atoms)
    unresolved_atoms = [
        cast(ob.OBAtom, atom.OBAtom)
        for atom in omol.atoms
        if has_unresolved_two_electron_center(atom.OBAtom)
    ]
    radical_sum = sum(
        get_unpaired_electron_count(cast(ob.OBAtom, atom.OBAtom))
        for atom in omol.atoms
        if not has_unresolved_two_electron_center(atom.OBAtom)
    )

    # A low-coordinate carbon can legitimately be represented as ``C-`` with
    # one radical electron. Fresh preparation defers this two-electron center;
    # resolve that specific one-unit charge/open-shell target here, where both
    # global budgets are available, instead of teaching generic charge cleanup
    # to partially consume unresolved markers.
    if (
        total_charge - formal_charge_sum == -1
        and len(unresolved_atoms) >= 1
        and total_radical_electrons - radical_sum == 1
    ):
        atom = unresolved_atoms[0]
        set_unresolved_two_electron_center(atom, False)
        set_unpaired_electron_count(atom, 1)
        atom.SetFormalCharge(-1)
        formal_charge_sum -= 1
        radical_sum += 1
        unresolved_atoms.remove(atom)

    if formal_charge_sum != total_charge:
        return False
    if any(
        cast(ob.OBBond, bond).GetBondOrder() > 5 or cast(ob.OBBond, bond).GetBondOrder() < 0
        for bond in ob.OBMolBondIter(cast(ob.OBMol, omol.OBMol))
    ):
        return False
    required_unpaired_electrons = total_radical_electrons - radical_sum
    if (
        required_unpaired_electrons < 0
        or required_unpaired_electrons % 2
        or required_unpaired_electrons > 2 * len(unresolved_atoms)
    ):
        return False

    triplet_centers = required_unpaired_electrons // 2
    for index, atom in enumerate(unresolved_atoms):
        set_unresolved_two_electron_center(atom, False)
        if index < triplet_centers:
            set_unpaired_electron_count(atom, 2)
            set_lone_pair_count(atom, 0)
        else:
            set_unpaired_electron_count(atom, 0)
            set_lone_pair_count(atom, 1)
    return True


def make_connections(
    omol: pybel.Molecule,
    extra_tolerance_angstrom: float = 0.15,
) -> tuple[pybel.Molecule, bool]:
    """Reconnect obvious donor/acceptor pairs before formal valence cleanup starts."""

    obmol = cast(ob.OBMol, omol.OBMol)
    hit = False
    while True:
        donate_atoms: List[int] = list(itertools.chain(*smarts.PREPROCESS_DONATE.findall(omol)))
        accept_atoms: List[int] = list(itertools.chain(*smarts.PREPROCESS_ACCEPT.findall(omol)))
        if not donate_atoms or not accept_atoms:
            break

        changed = False
        for donate_atom_id in donate_atoms:
            pairs = sorted(
                [
                    (donate_atom_id, accept_atom_id)
                    for accept_atom_id in accept_atoms
                    if accept_atom_id != donate_atom_id
                ],
                key=lambda x: cast(ob.OBAtom, obmol.GetAtom(x[0])).GetDistance(obmol.GetAtom(x[1])),
            )
            for pair_1, pair_2 in pairs:
                donate_atom = cast(ob.OBAtom, obmol.GetAtom(pair_1))
                accept_atom = cast(ob.OBAtom, obmol.GetAtom(pair_2))
                distance = cast(float, donate_atom.GetDistance(accept_atom))
                if (
                    distance
                    >= cast(
                        float,
                        ob.GetCovalentRad(donate_atom.GetAtomicNum())
                        + ob.GetCovalentRad(accept_atom.GetAtomicNum()),
                    )
                    + extra_tolerance_angstrom
                ):
                    continue

                bond = cast(Optional[ob.OBBond], obmol.GetBond(pair_1, pair_2))
                if bond is None:
                    obmol.AddBond(pair_1, pair_2, 1)
                    hit = True
                    changed = True
                    break
                if bond.GetBondOrder() == 0:
                    bond.SetBondOrder(1)
                    hit = True
                    changed = True
                    break
            if changed:
                break
        if not changed:
            break
    return omol, hit


def pre_clean(omol: pybel.Molecule) -> tuple[pybel.Molecule, bool]:
    """Apply cheap structural fixes that simplify the later heuristic stages."""

    obmol = cast(ob.OBMol, omol.OBMol)
    hit = False

    while res := smarts.PRE_CLEAN_HYPERVALENT.findall(omol):
        idxs_1 = cast(Tuple[int, int], res.pop(0))
        obbond = cast(ob.OBBond, obmol.GetBond(idxs_1[0], idxs_1[1]))
        obbond.SetBondOrder(obbond.GetBondOrder() - 1)
        hit = True

    while res := smarts.PRE_CLEAN_HYPER_PI_BOND.findall(omol):
        idxs_2 = cast(Tuple[int, int], res.pop(0))
        obbond = cast(ob.OBBond, obmol.GetBond(idxs_2[0], idxs_2[1]))
        obbond.SetBondOrder(1)
        hit = True

    while res := smarts.PRE_CLEAN_BCP_RING_5.findall(omol):
        idxs_3 = cast(Tuple[int, int, int, int, int], res.pop(0))
        bcp_n: Optional[int] = None
        bcp_c: Optional[int] = None
        for idx in idxs_3:
            indexs = set(idxs_3) - {idx}
            if all(cast(ob.OBBond, obmol.GetBond(idx, idx_3)) for idx_3 in indexs):
                if cast(ob.OBAtom, obmol.GetAtom(idx)).GetAtomicNum() == 7:
                    bcp_n = idx
                if cast(ob.OBAtom, obmol.GetAtom(idx)).GetAtomicNum() == 6:
                    bcp_c = idx
        if bcp_n is not None and bcp_c is not None:
            obmol.DeleteBond(cast(ob.OBBond, obmol.GetBond(bcp_n, bcp_c)))
            hit = True

    while res := smarts.PRE_CLEAN_BCP_RING_4.findall(omol):
        idxs_4 = cast(Tuple[int, int, int, int], res.pop(0))
        amine_n: Optional[int] = None
        butyl_c: Optional[int] = None
        for idx in idxs_4:
            indexs = set(idxs_4) - {idx}
            if all(obmol.GetBond(idx, idx_4) for idx_4 in indexs):
                if cast(ob.OBAtom, obmol.GetAtom(idx)).GetAtomicNum() == 7:
                    amine_n = idx
                if cast(ob.OBAtom, obmol.GetAtom(idx)).GetAtomicNum() == 6:
                    butyl_c = idx
        if amine_n is not None and butyl_c is not None:
            obmol.DeleteBond(obmol.GetBond(amine_n, butyl_c))
            hit = True

    while res := smarts.PRE_CLEAN_SI_O_F.findall(omol):
        idxs = cast(Tuple[int, int], res.pop(0))
        obmol.DeleteBond(obmol.GetBond(idxs[0], idxs[1]))
        hit = True

    return omol, hit


__all__ = ["make_connections", "pre_clean", "validate_omol"]

File: src/cpp/src/stages/preprocess.cpp
#include "molgr/stages/preprocess.h"

#include "molgr/stages/internal_helpers.h"

#include "molgr/utils/logger.h"
#include "molgr/utils/electrons.h"
#include "molgr/utils/smarts.h"
#include "molgr/utils/utils.h"

#include <openbabel/atom.h>
#include <openbabel/bond.h>
#include <openbabel/elements.h>
#include "molgr/compat/openbabel_iter.h"

#include <algorithm>

namespace molgr
{
    namespace reconstruct
    {
        using namespace OpenBabel;

        std::vector<int> GetFlatAtomList(OBMol &mol, molgr::smarts::PatternId pattern_id)
        {
            std::vector<int> atom_indices;
            auto matches = molgr::smarts::FindAll(mol, pattern_id);
            for (const auto &match : matches)
            {
                for (int idx : match)
                    atom_indices.push_back(idx);
            }
            return atom_indices;
        }

        // Electron bookkeeping: fixed atoms contribute only real unpaired
        // electrons to the radical budget. Resolve deferred centers deterministically
        // as triplet (2,0) or singlet (0,1); active lone pairs never count as spin.
        // Reject an impossible budget before mutating any center.
        bool ValidateOmol(OBMol &mol, int total_charge, int total_radical, bool emit_warnings)
        {
            int charge_sum = 0;
            int radical_sum = 0;
            std::vector<OpenBabel::OBAtom *> unresolved_atoms;

            FOR_ATOMS_OF_MOL(atom_iter, mol)
            {
                OpenBabel::OBAtom *atom = &(*atom_iter);
                if (molgr::utils::HasUnresolvedTwoElectronCenter(*atom))
                {
                    unresolved_atoms.push_back(atom);
                }
                charge_sum += atom->GetFormalCharge();
                if (!molgr::utils::HasUnresolvedTwoElectronCenter(*atom))
                {
                    radical_sum += molgr::utils::GetUnpairedElectronCount(*atom);
                }
            }

            // Resolve the specific open-shell anion form ``C-`` + one radical
            // only when both global budgets require it. Generic charge stages
            // intentionally never partially consume an unresolved marker.
            if (total_charge - charge_sum == -1 &&
                !unresolved_atoms.empty() &&
                total_radical - radical_sum == 1)
            {
                auto *atom = unresolved_atoms.front();
                molgr::utils::SetUnresolvedTwoElectronCenter(*atom, false);
                molgr::utils::SetUnpairedElectronCount(*atom, 1);
                molgr::utils::SetLonePairCount(*atom, 0);
                atom->SetFormalCharge(-1);
                --charge_sum;
                ++radical_sum;
                unresolved_atoms.erase(unresolved_atoms.begin());
            }

            if (charge_sum != total_charge)
            {
                if (emit_warnings)
                {
                    LOG_WARN("[Validate] Charge mismatch. Target: " << total_charge << ", Actual: " << charge_sum);
                }
                return false;
            }

            const int required_unpaired_electrons = total_radical - radical_sum;
            if (required_unpaired_electrons < 0 ||
                required_unpaired_electrons % 2 != 0 ||
                required_unpaired_electrons >
                    2 * static_cast<int>(unresolved_atoms.size()))
            {
                if (emit_warnings)
                {
                    LOG_WARN("[Validate] Radical budget cannot resolve "
                             << unresolved_atoms.size()
                             << " two-electron centers. Target: " << total_radical
                             << ", Fixed: " << radical_sum);
                }
                return false;
            }

            const int triplet_center_count = required_unpaired_electrons / 2;
            for (std::size_t index = 0; index < unresolved_atoms.size(); ++index)
            {
                auto &atom = *unresolved_atoms[index];
                molgr::utils::SetUnresolvedTwoElectronCenter(atom, false);
                if (static_cast<int>(index) < triplet_center_count)
                {
                    molgr::utils::SetUnpairedElectronCount(atom, 2);
                    molgr::utils::SetLonePairCount(atom, 0);
                }
                else
                {
                    molgr::utils::SetUnpairedElectronCount(atom, 0);
                    molgr::utils::SetLonePairCount(atom, 1);
                }
            }
            return true;
        }

        bool MakeConnections(OBMol &mol, double extra_tolerance_angstrom)
        {
            bool hit = false;
            while (true)
            {
                std::vector<int> donate_atoms = GetFlatAtomList(mol, molgr::smarts::PatternId::PREPROCESS_DONATE);
                std::vector<int> accept_atoms = GetFlatAtomList(mol, molgr::smarts::PatternId::PREPROCESS_ACCEPT);
                if (donate_atoms.empty() || accept_atoms.empty())
                {
                    break;
                }

                bool changed = false;
                for (int donor_idx : donate_atoms)
                {
                    OBAtom *donor = mol.GetAtom(donor_idx);
                    if (!donor)
                    {
                        continue;
                    }

                    std::vector<std::pair<int, int>> pairs;
                    for (int acceptor_idx : accept_atoms)
                    {
                        OBAtom *acceptor = mol.GetAtom(acceptor_idx);
                        if (acceptor && acceptor_idx != donor_idx)
                        {
                            pairs.push_back({donor_idx, acceptor_idx});
                        }
                    }

                    std::sort(pairs.begin(), pairs.end(), [&mol](const auto &a, const auto &b)
                              {
                        OBAtom *a1 = mol.GetAtom(a.first);
                        OBAtom *a2 = mol.GetAtom(a.second);
                        OBAtom *b1 = mol.GetAtom(b.first);
                        OBAtom *b2 = mol.GetAtom(b.second);
                        return a1->GetDistance(a2) < b1->GetDistance(b2); });

                    for (const auto &pair : pairs)
                    {
                        int p1 = pair.first;
                        int p2 = pair.second;
                        OBAtom *a1 = mol.GetAtom(p1);
                        OBAtom *a2 = mol.GetAtom(p2);
                        if (!a1 || !a2)
                        {
                            continue;
                        }
                        double dist = a1->GetDistance(a2);

                        double r1 = OBElements::GetCovalentRad(a1->GetAtomicNum());
                        double r2 = OBElements::GetCovalentRad(a2->GetAtomicNum());

                        if (dist >= (r1 + r2) + extra_tolerance_angstrom)
                        {
                            continue;
                        }

                        OBBond *bond = mol.GetBond(a1, a2);
                        if (!bond)
                        {
                            mol.AddBond(p1, p2, 1);
                            hit = true;
                            changed = true;
                            LOG_DEBUG("[MakeConnections] Add Bond " << p1 << "-" << p2);
                            break;
                        }
                        if (bond->GetBondOrder() == 0)
                        {
                            bond->SetBondOrder(1);
                            hit = true;
                            changed = true;
                            LOG_DEBUG("[MakeConnections] Set Bond Order 1 " << p1 << "-" << p2);
                            break;
                        }
                    }
                    if (changed)
                    {
                        break;
                    }
                }
                if (!changed)
                {
                    break;
                }
            }
            return hit;
        }

        bool PreClean(OBMol &mol)
        {
            bool hit = false;
            while (true)
            {
                auto matches1 = molgr::smarts::FindAll(mol, molgr::smarts::PatternId::PRE_CLEAN_HYPERVALENT);
                if (matches1.empty())
                {
                    break;
                }
                const auto &match = matches1.front();
                OBBond *bond = mol.GetBond(match[0], match[1]);
                if (bond)
                {
                    bond->SetBondOrder(bond->GetBondOrder() - 1);
                    hit = true;
                }
            }

            while (true)
            {
                auto matches2 = molgr::smarts::FindAll(
                    mol,
                    molgr::smarts::PatternId::PRE_CLEAN_HYPER_PI_BOND);
                if (matches2.empty())
                {
                    break;
                }
                const auto &match = matches2.front();
                OBBond *bond = mol.GetBond(match[0], match[1]);
                if (bond)
                {
                    bond->SetBondOrder(1);
                    hit = true;
                }
            }

            while (true)
            {
                auto matches3 = molgr::smarts::FindAll(mol, molgr::smarts::PatternId::PRE_CLEAN_BCP_RING_5);
                if (matches3.empty())
                {
                    break;
                }
                const auto &idxs = matches3.front();
                int n_idx = -1, c_idx = -1;
                for (int idx : idxs)
                {
                    OBAtom *atom = mol.GetAtom(idx);
                    int internal_degree = 0;
                    for (int other : idxs)
                        if (idx != other && mol.GetBond(idx, other))
                            internal_degree++;

                    if (internal_degree >= 3)
                    {
                        if (atom->GetAtomicNum() == 7)
                            n_idx = idx;
                        if (atom->GetAtomicNum() == 6)
                            c_idx = idx;
                    }
                }
                if (n_idx != -1 && c_idx != -1)
                {
                    OBBond *bond = mol.GetBond(n_idx, c_idx);
                    if (bond)
                    {
                        mol.DeleteBond(bond);
                        hit = true;
                    }
                }
            }

            while (true)
            {
                auto matches4 = molgr::smarts::FindAll(mol, molgr::smarts::PatternId::PRE_CLEAN_BCP_RING_4);
                if (matches4.empty())
                {
                    break;
                }
                const auto &idxs = matches4.front();
                int amine_n = -1, butyl_c = -1;
                for (int idx : idxs)
                {
                    OBAtom *atom = mol.GetAtom(idx);
                    int internal_degree = 0;
                    for (int other : idxs)
                        if (idx != other && mol.GetBond(idx, other))
                            internal_degree++;
                    if (internal_degree >= 2)
                    {
                        if (atom->GetAtomicNum() == 7)
                            amine_n = idx;
                        if (atom->GetAtomicNum() == 6)
                            butyl_c = idx;
                    }
                }
                if (amine_n != -1 && butyl_c != -1)
                {
                    OBBond *bond = mol.GetBond(amine_n, butyl_c);
                    if (bond)
                    {
                        mol.DeleteBond(bond);
                        hit = true;
                    }
                }
            }

            while (true)
            {
                auto matches5 = molgr::smarts::FindAll(mol, molgr::smarts::PatternId::PRE_CLEAN_SI_O_F);
                if (matches5.empty())
                {
                    break;
                }
                const auto &match = matches5.front();
                OBBond *bond = mol.GetBond(match[0], match[1]);
                if (bond)
                {
                    mol.DeleteBond(bond);
                    hit = true;
                }
            }
            return hit;
        }
    }
}
